diff --git a/index.js b/index.js
--- a/index.js
+++ b/index.js
@@ -132,7 +132,7 @@
     const parts = header.trim().split(/,/);
     for (const part of parts) {
         const [k, v] = part.split(/=/, 2);
-        cc[k.trim()] = v === undefined ? true : v.trim().replace(/^"|"$/g, '');
+        cc[k.trim().toLowerCase()] = v === undefined ? true : v.trim().replace(/^"|"$/g, '');
     }
 
     return cc;
@@ -395,8 +395,7 @@
     evaluateRequest(req) {
         this._assertRequestHasHeaders(req);
 
-        // In all circumstances, a cache MUST NOT ignore the must-revalidate directive
-        if (this._rescc['must-revalidate']) {
+        if (!this._allowsReuse()) {
             return this._evaluateRequestMissResult(req);
         }
 
@@ -427,7 +426,7 @@
             // If a value is present, then the client is willing to accept a response that has
             // exceeded its freshness lifetime by no more than the specified number of seconds
             const allowsStaleWithoutRevalidation = 'max-stale' in requestCC &&
-                (true === requestCC['max-stale'] || requestCC['max-stale'] > this.age() - this.maxAge());
+                (true === requestCC['max-stale'] || requestCC['max-stale'] >= this.age() - this.maxAge());
 
             if (allowsStaleWithoutRevalidation) {
                 return this._evaluateRequestHitResult(undefined);
@@ -640,9 +639,6 @@
         }
 
         if (this._isShared) {
-            if (this._rescc['proxy-revalidate']) {
-                return 0;
-            }
             // if a response includes the s-maxage directive, a shared cache recipient MUST ignore the Expires field.
             if (this._rescc['s-maxage']) {
                 return toNumberOrZero(this._rescc['s-maxage']);
@@ -702,11 +698,31 @@
         return this.maxAge() <= this.age();
     }
 
+    _allowsReuse() {
+        if (!this.storable() || this._rescc['no-cache']) {
+            return false;
+        }
+        if (!this.stale()) {
+            return true;
+        }
+        return !(
+            this._rescc['must-revalidate'] ||
+            (this._isShared && (
+                this._rescc['proxy-revalidate'] ||
+                this._rescc['s-maxage'] ||
+                (this._resHeaders['set-cookie'] &&
+                    !this._rescc.public &&
+                    !this._rescc.immutable)
+            ))
+        );
+    }
+
     /**
      * @returns {boolean} `true` if `stale-if-error` condition allows use of a stale response.
      */
     _useStaleIfError() {
-        return this.maxAge() + toNumberOrZero(this._rescc['stale-if-error']) > this.age();
+        const sie = toNumberOrZero(this._rescc['stale-if-error']);
+        return sie > 0 && this._allowsReuse() && this.maxAge() + sie > this.age();
     }
 
     /** See `evaluateRequest()` for a more complete solution
@@ -714,7 +730,7 @@
      */
     useStaleWhileRevalidate() {
         const swr = toNumberOrZero(this._rescc['stale-while-revalidate']);
-        return swr > 0 && this.maxAge() + swr > this.age();
+        return swr > 0 && this._allowsReuse() && this.maxAge() + swr > this.age();
     }
 
     /**
@@ -855,7 +871,7 @@
     revalidatedPolicy(request, response) {
         this._assertRequestHasHeaders(request);
 
-        if (this._useStaleIfError() && isErrorResponse(response)) {
+        if (this._requestMatches(request, true) && this._useStaleIfError() && isErrorResponse(response)) {
           return {
               policy: this,
               modified: false,
